--- 12.tiny.c	2017-11-09 02:57:43.655936083 +0000
+++ 12.35.c	2017-11-09 02:57:43.655936083 +0000
@@ -1,6 +1,8 @@
 /*
- * tiny.c - A simple, iterative HTTP/1.0 Web server that uses the
+ * 12.35.c - A simple, iterative HTTP/1.0 Web server that uses the
  *     GET method to serve static and dynamic content.
+ *
+ *     concurrent server in multi process way.
  */
 #include "csapp.h"
 
@@ -35,8 +37,14 @@
     Getnameinfo((SA *) &clientaddr, clientlen, hostname, MAXLINE,
         port, MAXLINE, 0);
     printf("Accepted connection from (%s, %s)\n", hostname, port);
-    doit(connfd);                                             //line:netp:tiny:doit
-    Close(connfd);                                            //line:netp:tiny:close
+
+    if (Fork() == 0) {
+      Close(listenfd);
+      doit(connfd);                                             //line:netp:tiny:doit
+      Close(connfd);                                            //line:netp:tiny:close
+      exit(0);
+    }
+    Close(connfd);
   }
 }
 
