--- tiny.origin.c	2017-11-09 02:57:43.651936112 +0000
+++ tiny.9.c	2017-11-09 02:57:43.651936112 +0000
@@ -152,12 +152,12 @@
   printf("Response headers:\n");
   printf("%s", buf);
 
-  /* Send response body to client */
   srcfd = Open(filename, O_RDONLY, 0);    //line:netp:servestatic:open
-  srcp = Mmap(0, filesize, PROT_READ, MAP_PRIVATE, srcfd, 0);//line:netp:servestatic:mmap
+  srcp = (char*)Malloc(filesize);
+  Rio_readn(srcfd, srcp, filesize);
   Close(srcfd);                           //line:netp:servestatic:close
   Rio_writen(fd, srcp, filesize);         //line:netp:servestatic:write
-  Munmap(srcp, filesize);                 //line:netp:servestatic:munmap
+  free(srcp);
 }
 
 /*
