--- ../vm/memlib.c	2017-11-09 02:57:43.679935907 +0000
+++ memlib.c	2017-11-09 02:57:43.675935936 +0000
@@ -15,23 +15,18 @@
 #include "csapp.h"
 #include "memlib.h"
 
-#define MAX_HEAP (20*(1<<20))  /* 20 MB */
-
-
 /* $begin memlib */
 /* Private global variables */
 static char *mem_heap;     /* Points to first byte of heap */
 static char *mem_brk;      /* Points to last byte of heap plus 1 */
-static char *mem_max_addr; /* Max legal heap addr plus 1*/
 
 /*
  * mem_init - Initialize the memory system model
  */
 void mem_init(void)
 {
-  mem_heap = (char *)Malloc(MAX_HEAP);
+  mem_heap = (char *)sbrk(0);
   mem_brk = (char *)mem_heap;
-  mem_max_addr = (char *)(mem_heap + MAX_HEAP);
 }
 
 /*
@@ -43,7 +38,7 @@
 {
   char *old_brk = mem_brk;
 
-  if ( (incr < 0) || ((mem_brk + incr) > mem_max_addr)) {
+  if ( (incr < 0) || ((mem_brk = sbrk(incr)) == (void *)-1)) {
     errno = ENOMEM;
     fprintf(stderr, "ERROR: mem_sbrk failed. Ran out of memory...\n");
     return (void *)-1;
