--- adder.c	2017-05-18 14:26:13.191017029 +0800
+++ head-adder.c	2017-05-26 15:44:54.912608094 +0800
@@ -1,10 +1,10 @@
 /*
- * adder.c - a minimal CGI program that adds two numbers together
+ * head-adder.c - a minimal CGI program that adds two numbers together
  */
 #include "../csapp.h"
 
 int main(void) {
-  char *buf, *p;
+  char *buf, *p, *method;
   char arg1[MAXLINE], arg2[MAXLINE], content[MAXLINE];
   int n1=0, n2=0;
 
@@ -18,6 +18,8 @@
     n2 = atoi(arg2);
   }
 
+  method = getenv("REQUEST_METHOD");
+
   /* Make the response body */
   sprintf(content, "Welcome to add.com: ");
   sprintf(content, "%sTHE Internet addition portal.\r\n<p>", content);
@@ -29,7 +31,10 @@
   printf("Connection: close\r\n");
   printf("Content-length: %d\r\n", (int)strlen(content));
   printf("Content-type: text/html\r\n\r\n");
-  printf("%s", content);
+
+  if (strcasecmp(method, "HEAD") != 0)
+    printf("%s", content);
+
   fflush(stdout);
 
   exit(0);
